### PR TITLE
Add research output content model and data provider

### PR DESCRIPTION
**Pull request by lennym**
TODO

- [x] instantiate data provider with correct (preview) token
- [ ] handle `usageNotes` pseudo-rich text
- [x] update method on data provider
- [x] data provider unit tests
- [x] working group filters
- [x] text and keyword search
- [x] check for unpublished changes when fetching published ROs
- [x] handle created timestamp offsets in squidex tests
- [x] contact emails
- [x] handle stale cache when editing a published output

### CODE REVIEW

**Review by gabiayako — COMMENTED** [review]
I still need to look at some files, but I am already submitting some comments before I go to have lunch :)

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 1]
```diff
@@ -0,0 +1,1541 @@
...
+    jest.useRealTimers();
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  describe('fetch', () => {
+    test('can call from mock server', async () => {
+      await researchOutputDataProviderMockGraphql.fetch({});
```
[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:68]: Yes, getting the graphql mock client to return meaningful data with multiple nested self-referential models was somewhere between very hard and impossible, so speaking to @peterstarling we agreed to simply check that the query completes without error as a basic check that the graphql query is well-formed and matches the schema.

File: packages/contentful/migrations/crn/researchOutputs/20230626131700-research-outputs.js [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+  researchOutputs
+    .createField('labCatalogNumber')
+    .name('Lab Catalog Number')
+    .type('Symbol')
+    .localized(false)
+    .required(false)
+    .validations([
+      {
+        regexp: {
+          pattern: '^(doi:)?d{2}.d{4}.*$',
```
[the remark this answers is not in the document]
Comment (lennym) [reply packages/contentful/migrations/crn/researchOutputs/20230626131700-research-outputs.js:719]: No, this was 100% a copy-paste oversight.

**Review by gabiayako — COMMENTED** [review]
I still haven't finished it 😅 I'm reviewing while CI/migration run in my task, so I'm coming back time to time to review it, that's why it's taking so long 😝 

I'm submitting more comments for now, but they're very minor, it's looking really good 😄

File: apps/crn-server/test/integration/common/research-outputs.integration-test.ts [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        });
+      });
+
+      test('can list draft research outputs for a team you are a member of', async () => {
+        await retryable(async () => {
+          await supertest(app)
+            .get(`/research-outputs?teamId=${nonPmTeam.id}&status=draft`)
+            .expect(200);
+        });
+      });
```
[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/integration/common/research-outputs.integration-test.ts:327]: Got to be _absolutely_ sure it's working...

File: apps/crn-server/test/integration/common/research-outputs.integration-test.ts [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+
+      test('can list draft research outputs for a working group you are a member of', async () => {
+        await retryable(async () => {
+          await supertest(app)
+            .get(
+              `/research-outputs?workingGroupId=${memberWorkingGroup.id}&status=draft`,
+            )
+            .expect(200);
+        });
+      });
```
[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/integration/common/research-outputs.integration-test.ts:502]: I've deleted these tests because it turns out that listing published outputs for a particular team is not an actual use case.

File: packages/contentful/migrations/crn/researchOutputs/20230626131700-research-outputs.js [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+  researchOutputs
+    .createField('rrid')
+    .name('Identifier (RRID)')
+    .type('Symbol')
+    .localized(false)
+    .required(false)
+    .validations([
+      {
+        regexp: {
+          pattern: '^RRID:[a-zA-Z]+.+$',
```
[the remark this answers is not in the document]
Comment (lennym) [reply packages/contentful/migrations/crn/researchOutputs/20230626131700-research-outputs.js:668]: I would if I could find literally any mention of that in any documentation... 😬

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 6]
```diff
@@ -0,0 +1,1541 @@
...
+    test('throws a not found error if query does not return a result', async () => {
+      contentfulGraphqlClientMock.request.mockResolvedValue({
+        researchOutputs: null,
+      });
+      await expect(researchOutputDataProvider.fetchById('1')).rejects.toThrow(
+        'Not Found',
+      );
+    });
+
+    test('looks for published version first, and then draft if there is no published version', async () => {
```
Comment (peterstarling) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:253]: As I understand this is done because our API does not actually behave differently whether you ask for a draft or published RO? In which case I wonder why you went with this approach instead of simply asking for the resource with  the preview flag set to true right from the start? If you were switching between the different tokens (as suggested further up) then that would make sense but at this stage i would say it's unnecessary

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 7]
```diff
@@ -0,0 +1,1541 @@
...
+      const researchOutputs = getContentfulResearchOutputGraphqlResponse();
+      contentfulGraphqlClientMock.request.mockResolvedValueOnce({
```
Comment (peterstarling) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:255]: I'd consider using the .when mock here but i'll leave it up to you to decide

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 8]
```diff
@@ -0,0 +1,1541 @@
...
+        researchOutputs,
+      });
+
+      const result = await researchOutputDataProvider.fetchById('1');
+
+      expect(result!.relatedResearch).toEqual([]);
+    });
+
+    test('Should combine inbound and outbound related research links', async () => {
+      const researchOutputs = getContentfulResearchOutputGraphqlResponse();
```
Comment (peterstarling) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:400]: right now you're depending on the fixture for this test, i like to be explicit when testing different edge cases, even if the fixture data already works for us

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 9]
```diff
@@ -0,0 +1,1541 @@
...
+
+    test('sets external authors ids to contentful references', async () => {
+      await researchOutputDataProvider.create(
+        {
+          ...getResearchOutputCreateDataObject(),
+          authors: [{ userId: 'user-1' }, { externalAuthorId: 'ea-1' }],
+        },
+        { publish: false },
+      );
+      const args = environmentMock.createEntry.mock.lastCall?.[1];
```
Comment (peterstarling) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:820]: no way to use  .toBeCalledWith instead?

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 10]
```diff
@@ -0,0 +1,1541 @@
...
+    });
+
+    test('can update an output', async () => {
+      await researchOutputDataProvider.update(
+        'a',
+        getResearchOutputUpdateDataObject(),
+        { publish: false },
+      );
+
+      expect(patch).toHaveBeenCalled();
```
Comment (peterstarling) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:1188]: you might want to test what it was called with cause i don't think the 'a' argument is being tested anywhere?

**Review by peterstarling — CHANGES_REQUESTED** [review]

File: .github/workflows/reusable-test.yml [hunk 11]
```diff
@@ -160,20 +170,29 @@ jobs:
         shell: bash
         run: |
           yarn workspace @asap-hub/contentful space:create-environment --timeout 600000
+          yarn contentful:migration:run:crn
```
[the remark this answers is not in the document]
Comment (lennym) [reply .github/workflows/reusable-test.yml:174]: If a PR (such as this one) includes migrations then they need to run before any integration tests because the environment clones from Development which won't have those tests.

In this case, without this line then all the integration tests for research outputs fail because research outputs don't exist.

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts (hunk 6 above)
Comment (lennym) [reply apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:253]: If an ASAP admin makes a change to a published output in the CMS, but then _doesn't_ publish the changes then this makes sure that the published version is still returned if that output is requested via the hub until such a time as they publish.

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 12]
```diff
@@ -0,0 +1,1611 @@
...
+                    linkType: 'Entry',
+                    id: 'team-2',
+                  },
+                },
+              ],
+            },
+          }),
+        }),
+      );
+    });
```
Comment (gabiayako) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:963]: this is also duplicated 😛

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 13]
```diff
@@ -0,0 +1,1611 @@
...
+              sys: {
+                type: 'Link',
+                linkType: 'Entry',
+                id: 'team-2',
+              },
+            },
+          ],
+        }),
+      );
+    });
```
Comment (gabiayako) [inline apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:1410]: This is also duplicated

File: apps/crn-server/src/data-providers/contentful/research-output.data-provider.ts [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+      FetchResearchOutputByIdQuery,
+      FetchResearchOutputByIdQueryVariables
+    >(FETCH_RESEARCH_OUTPUT_BY_ID, { id, preview: false });
+  }
+
+  async fetchById(id: string): Promise<ResearchOutputDataObject | null> {
+    const { researchOutputs } = await this.fetchOutputById(id);
+
+    if (!researchOutputs) {
+      const { researchOutputs: draftResearchoutput } =
```
Comment (gabiayako) [inline apps/crn-server/src/data-providers/contentful/research-output.data-provider.ts:151]:
```suggestion
      const { researchOutputs: draftResearchOutput } =
```

**Review by gabiayako — APPROVED** [review]
🎉

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts [hunk 15]
```diff
@@ -0,0 +1,1541 @@
...
+            expect.objectContaining({
+              limit: 13,
+              skip: 7,
+            }),
+          );
+        });
+
+        test('Should pass the parameters to fetch drafts as expected', async () => {
+          await researchOutputDataProvider.fetch({
+            filter: {
```
[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:140]: The behaviour is currently undefined because those parameters are [tightly coupled at the router level](https://github.com/yldio/asap-hub/blob/master/apps/crn-server/src/routes/research-output.route.ts#L34).

Having said that, the behaviour of contentful if the draft filter is not set, but the preview API _is_ used is probably problematic (because it will surface unpublished changes to published outputs) so we might want to enforce that same coupling at a data provider level.

I'm not sure if there is any use case where we fetch draft _and_ published outputs in the same request though.

[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:140]: So there is a use case (uniqueness checks) and it _is_ problematic.

File: apps/crn-server/test/integration/common/research-tags.integration-test.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    loggedInUser = await fixtures.createUser(getUserFixture());
+    app = AppHelper(() => loggedInUser);
+  });
+
+  afterAll(async () => {
+    await fixtures.teardown();
+  });
+
+  test('can fetch a list of research tags', async () => {
+    const response = await supertest(app).get('/research-tags').expect(200);
```
[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/integration/common/research-tags.integration-test.ts:25]: We _could_ do that, but it's a lot of setup to create all the different types of keyword for marginal value.

I just ran the keyword migration once on the environment, which puts all ~850 keywords from Squidex into contentful, and then work on the assumption that they're going to be there (and nothing especially bad happens if they're not).

File: apps/crn-server/test/integration/helpers/app.ts [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import express, { Express } from 'express';
+import { appFactory } from '../../../src/app';
+import { decodeTokenFactory } from '@asap-hub/server-common';
+import { UserFixture } from '../fixtures';
+
+jest.mock('../../../src/config', () => ({
```
[the remark this answers is not in the document]
Comment (lennym) [reply apps/crn-server/test/integration/helpers/app.ts:6]: The config is [imported directly into the feature flag middleware](https://github.com/yldio/asap-hub/blob/master/apps/crn-server/src/middleware/feature-flag.ts#L2) without any DI, so I'm not sure that's true.

Is this _not_ a test file? Unless you mean the top level test suites? We _could_ move it there, but it felt unnecessarily repetitive to have the same boilerplate in the top of every file.

File: apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts (hunk 9 above)
Comment (lennym) [reply apps/crn-server/test/data-providers/contentful/research-output.data-provider.test.ts:922]: No, it was the several layers of nested objects, which meant a single use of `expect.objectContaining` wasn't sufficient, but replacing with several instances inside one another was fine.
